src/function/table/arrow.cpp

```diff
@@ -275,15 +252,77 @@ static void ArrowScanFunction(ClientContext &context, const FunctionData *bind_d
 			throw std::runtime_error("Unsupported type " + output.data[col_idx].GetType().ToString());
 		}
 	}
+}
+
+static void ArrowScanFunction(ClientContext &context, const FunctionData *bind_data,
+                              FunctionOperatorData *operator_state, DataChunk *input, DataChunk &output) {
+	auto &data = (ArrowScanFunctionData &)*bind_data;
+	auto &state = (ArrowScanState &)*operator_state;
+	//! have we run out of data on the current chunk? move to next one
+	if (state.chunk_offset >= (idx_t)state.chunk->length) {
+		state.chunk_offset = 0;
+		state.ReleaseArray();
+		delete state.chunk;
+		if (data.is_consumed) {
+			//! No more chunks
+			return;
+		}
+		auto current_chunk = new ArrowArray();
+		if (data.stream->get_next(data.stream, current_chunk)) {
```

Comment:
There is a race condition on data.is_consumed. It might not matter but data.stream->get_next can be called multiple times after it has been emptied by different threads.